feat: Add Langchain integration (#120)
SAP/ai-sdk-js: packages/langchain/README.md

  ['system', 'Answer the following in {language}:'],
  ['user', '{text}']
]);
const parser = new StringOutputParser();
const llmChain = promptTemplate.pipe(client).pipe(parser);
const response = await llmChain.invoke({
  language: 'german',
  text: 'What is the capital of France?'
});
```

### Embedding Client
Embedding clients allow embedding either text or documents (represented as arrays of strings).

#### Embed Text
```ts
const embeddedText = await embeddingClient.embedQuery(
  'Paris is the capital of France.'
);
```

#### Embed Documents
```ts
const embeddedDocument = await embeddingClient.embedDocuments([
  'Page 1: Paris is the capital of France.',
  'Page 2: It is a beautiful city.'
]);
```

## Support, Feedback, Contribution
This project is open to feature requests/suggestions, bug reports etc. via [GitHub issues](https://github.com/SAP/ai-sdk-js/issues).

Contribution and feedback are encouraged and always welcome. For more information about how to contribute, the project structure, as well as additional contribution information, see our [Contribution Guidelines](https://github.com/SAP/ai-sdk-js/blob/main/CONTRIBUTING.md).

## License

> The SAP Cloud SDK for AI is released under the [Apache License Version 2.0.](http://www.apache.org/licenses/)

Reviewer: ```suggestion
The SAP Cloud SDK for AI is released under the [Apache License Version 2.0.](http://www.apache.org/licenses/).
```

# @sap-ai-sdk/langchain
This package provides LangChain model clients, built on top of the `@sap-ai-sdk`.

## Table of Contents
1. [Installation](#installation)
2. [Pre-requisites](#pre-requisites)
3. [Usage](#usage)
    - [Client Initialization](#client-initialization)
    - [Chat Clients](#chat-clients)
    - [Embedding Clients](#embedding-clients)
4. [Support, Feedback, Contribution](#support-feedback-contribution)

> 5. [License](#license)

Reviewer: @KavithaSiva

## Installation
```
$ npm install @sap-ai-sdk/langchain
```

## Pre-requisites
- [Enable the AI Core service in SAP BTP](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/initial-setup).
- Bind the service to your application.
- Ensure the project is configured with Node.js v20 or higher, along with native ESM support.
- For testing your application locally:
  - Download a service key for your AI Core service instance.
  - Create a `.env` file in the sample-code directory.
  - Add an entry `AICORE_SERVICE_KEY='<content-of-service-key>'`.

## Usage

> This package provides both chat and embedding clients, currently supporting Azure OpenAI.

Reviewer: ```suggestion
This package provides chat and embedding clients, currently supporting Azure OpenAI.
```

All clients comply with [LangChain's interface](https://js.langchain.com/docs/introduction).

### Client Initialization

> To initialize a client, you only need to provide the model name:

Reviewer 2: ```suggestion
To initialize a client, provide the model name:
```

```ts
import { AzureOpenAiChatClient } from '@sap-ai-sdk/langchain';
const chatClient = new AzureOpenAiChatClient({ modelName: 'gpt-4o' });
```

In addition to the default parameters of the model vendor (e.g. OpenAI) and LangChain, there are also SDK-specific parameters, which you can use to narrow down the search for the model you want to use:

```ts
const chatClient = new AzureOpenAiChatClient({
  modelName: 'gpt-4o',
  modelVersion: '24-07-2021',
  resourceGroup: 'my-resource-group'
});
```

> ### Chat Clients

Reviewer: ```suggestion
### Chat Client
```

> ### Chat Clients

Reviewer: [q] Should it be singular?

The chat clients allow you to interact with Azure OpenAI chat models, accessible via SAP Gen AI Hub. 
To invoke the client, you only have a to pass a prompt.

> #### Simple Example

Reviewer: [pp] I would leave out this heading here, put a colon before the last sentence 
`To invoke the client, you only have to pass a prompt: `

```ts
const response = await chatClient.invoke("What's the capital of France?");
```

#### Advanced Example with Templating and Output Parsing
```ts
import { AzureOpenAiChatClient } from '@sap-ai-sdk/langchain';
import { StringOutputParser } from '@langchain/core/output_parsers';
import { ChatPromptTemplate } from '@langchain/core/prompts';

const client = new AzureOpenAiChatClient({ modelName: 'gpt-35-turbo' });
const promptTemplate = ChatPromptTemplate.fromMessages([
  ['system', 'Answer the following in {language}:'],
  ['user', '{text}']
]);
const parser = new StringOutputParser();
const llmChain = promptTemplate.pipe(client).pipe(parser);
const response = await llmChain.invoke({
  language: 'german',
  text: 'What is the capital of France?'
});
```

### Embedding Clients

> Embedding clients allow embedding either text or documents (represented as arrays of strings).

Author: The only difference is the name of the client, the initialization is exactly the same, therefore I wouldn't add an additional section.

> Embedding clients allow embedding either text or documents (represented as arrays of strings).

Reviewer 3: while that is true, users don't know that and in the below code example can only guess what `embeddingClient` refers to. We also can't assume that all consumers of embedding functionality are also consumers of chat functionality. Thus we need to have something IMO, either a small piece of code or text

> # @sap-ai-sdk/LangChain

Reviewer 2: [req] nope!

# @sap-ai-sdk/langchain

> This package contains Langchain compliant models, based on the @sap-ai-sdk clients.

Reviewer 2: [req] It is called "LangChain" (pascal case). Please make sure to always refer to it by its correct name.

> ### Installation

Author: I believe the license etc. has to be in here still?
The packages are released as (somewhat) independent parts on npm, or am I missing something?

```
$ npm install @sap-ai-sdk/langchain

$ npm install @langchain/openai // if you want to use OpenAI models
```

## Pre-requisites

- [Enable the AI Core service in BTP](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/initial-setup).

> - Project configured with Node.js v20 or higher and native ESM support enabled.

Reviewer 2: [req]All other points are in imperative. I don't care whether you use the imperative or past tense for this, but it should be consistent.

- For testing your application locally:
  - Download a service key for your AI Core service instance.

>   - Create a `.env` file in the sample-code directory.

Reviewer 2: [q] How is that related to "testing your application locally"? Please note, that those are the instruction on npm. Seems a bit confusing to me and rather related to our sample code / E2E tests.

Author: I figured you have to follow the same steps when using this library and then testing your app.
I can see how this might be redundant though, as this could then be said for any library ever. 🤷

  - Add an entry `AICORE_SERVICE_KEY='<content-of-service-key>'`.

## Usage

> All client's comply with [Langchain's interface](https://python.langchain.com/v0.2/api_reference/openai/chat_models/langchain_openai.chat_models.azure.AzureChatOpenAI.html#langchain_openai.chat_models.azure.AzureChatOpenAI), therefore you should be able to use them as per usual.

Reviewer 2: [pp] "you should be able to use them as usual" seems like:
1. we are unsure whether it works, indicating low quality on our side.
2. doppelt-gemoppelt. If it complies with langchain it means you can use it as expected, so why mention it explicitly. I would omit the part after the comma.

> All client's comply with [Langchain's interface](https://python.langchain.com/v0.2/api_reference/openai/chat_models/langchain_openai.chat_models.azure.AzureChatOpenAI.html#langchain_openai.chat_models.azure.AzureChatOpenAI), therefore you should be able to use them as per usual.

Reviewer 2: [q/req] Why are we linking to the python API foe AzureOpenAI for Chat here. Seems oddly specific for the whole LangChain package and irrelevant as it is python. I think it would be more useful to link to JS (=>https://js.langchain.com/docs/introduction) and maybe not link to a specifc API version, but a general explanation of the concept.

> The only difference is in the initialization of the client, where you have th option to pass either:

Reviewer 2: [pp] => "the only difference" implies that we are only almost compliant with the langchain interface and now we explain a workaround, because we do something special. To my understanding this is not the case.

Consider something more neutral in this direction:
```suggestion
To initialize the client pass either:
```

Still, I would even prefer it to lead with the main use case of passing the model name, so that people can find this  as a TLDR and then follow up with additional information like the fact that you can pass a model version and resource group optionally or instead pass the deployment ID, if you want to discover it on your own.

```ts
    modelName: string,
    modelVersion?: string,
    resourceGroup?: string,
    ...others
```

or

```ts
    deploymentId: string,
    resourceGroup?: string
    ...others
```

> If you pass API keys they are ignored, since you're not inteded to call the vendor's endpoints directly.

Reviewer 2: [req] I think you cannot say "you are not intended" => "du bist nicht beabsichtigt". "it is not intended" or "you are not supposed to" is correct.
Also I would keep "are" ignored as I remember from the technical writing course to avoid future tense ;)

> Instead, the credentials in the binding are used to call SAP's LLM Proxy.

Reviewer 2: [q] is this the official name?

Author: I wasn't sure what the name of the component we call with the credentials is, do you have a better suggestion? 😬

Reviewer 4: ```suggestion
Instead, credentials in the binding are required to access SAP AI Core capabilities, specifically for consuming LLMs via the SAP Generative AI Hub.
```

### OpenAI

We offer two types of clients for OpenAI models.

> Currenty these are chat and embedding models.

Reviewer 2: [pp] agreed with Deeksha +:
```suggestion
```

All clients assume the same set of default values as [Langchain's default OpenAI client](https://www.npmjs.com/package/@langchain/openai) does.

#### Chat

There are two common APIs, `.invoke()` for simple text completion and `.generate()` for chat completion.
You can also combine them with the usual Langchain functionality, e.g. prompt templates.

A simple text completion might look like:

##### Initialization

```ts
import { OpenAiChatClient } from '@sap-ai-sdk/langchain';
const chatClient = new OpenAiChatClient({ modelName: 'gpt-4o' });
```

##### Usage

```ts
const response = await chatClient.invoke("What's the capital of France?'");
```

A chat completion example might be:

```ts
const response = await chatClient.generate([
  [new SystemMessage('You are an IT support agent answering questions.')],
  [new HumanMessage('Why is my internet not working?')]
]);

> ```

Author: Yeah that's a good point, I tried to align the documentation to the LLM-Common documentation to some degree, but I suppose the only difference between the different clients will be the name, not their interface.
